src/models.jl

```diff
@@ -0,0 +1,124 @@
+mutable struct Model
+    metadata::ModelMetadata
+    configuration::ModelConfiguration
+    boundary_conditions::BoundaryConditions
+    constants::PlanetaryConstants
+    eos::EquationOfState
+    grid::Grid
+    velocities::VelocityFields
+    tracers::TracerFields
+    pressures::PressureFields
+    G::SourceTerms
+    Gp::SourceTerms
+    forcings::ForcingFields
+    forcing::Forcing
+    stepper_tmp::StepperTemporaryFields
+    operator_tmp::OperatorTemporaryFields
+    ssp  # ::SpectralSolverParameters or ::SpectralSolverParametersGPU
+    clock::Clock
+    output_writers::Array{OutputWriter,1}
+    diagnostics::Array{Diagnostic,1}
+end
+
+"""
+    Model(; kwargs...)
+
+Construct an `Oceananigans.jl` model. The keyword arguments are:
+
+          N (tuple) : model resolution in (x, y, z)
+          L (tuple) : model domain in (x, y, z)
+         dt (float) : time step
+ start_time (float) : start time for the simulation
+  time_step (float) : ?
+         arch (sym) : architecture (:cpu or :gpu)
+  float_type (type) : floating point type for model data (typically Float32 or Float64)
+          constants : planetary constants (?)
+                eos : equation of state to infer density from temperature and salinity
+            forcing : forcing functions for (u, v, w, T, S)
+boundary_conditions : boundary conditions
+     output_writers : output writer
+        diagonstics : diagnostics
+
+        ... and more.
+"""
+function Model(;
+    # Model resolution and domain size
+             N = (32, 1, 32),
+             L = (1.0, 1.0, 1.0),
```

Comment:
Don't think we should be specifying defaults for `N` or `L`. I think the user should be specifying this at a minimum.

Reply:
Why? I think it's sorta nice for sanity that a user can say `model = Model()` and it works and the user can already start to experiment. We can even made a dumb model that works if you just write the lines 

```julia
model = Model()
step!(model)
plot(model)
```

Three lines to get started? Pretty nice?